### PR TITLE
ASoC: Intel: sof_sdw: add multi codec dai per codec support 

### PR DESCRIPTION
**Pull request by bardliao**
The existing sdw_sof machine driver assumes one codec dai will be connected per codec and create a dai link per codec.
However, the rt712 codec supports jack and amp on different dais. The PR removed this restriction and can create two different dai links for rt712 jack and amp.

### CODE REVIEW

**Comment by greg-intel** [comment]
SOFCI TEST

File: sound/soc/intel/boards/sof_sdw.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: do we really want to do this? This would require us to have codec-specific stuff in the topology, no?

File: sound/soc/intel/boards/sof_sdw_common.h
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw_common.h]: where does this come from?

File: sound/soc/intel/boards/sof_sdw_common.h
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw_common.h]: this is a bit misleading, SoundWire DAIs are not bi-directional.

**Review by plbossart — COMMENTED** [review]
Not getting the introduction of the codec_name in the dai link name @bardliao. I am not sure why we would do this because there is currently no reference at all to the codecs in the topology

File: sound/soc/intel/boards/sof_sdw_common.h
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_common.h]: > this is a bit misleading, SoundWire DAIs are not bi-directional.

We have the same thought @plbossart My first version is to change it to `const int direction[SOF_SDW_CODEC_MAX_DAI_NUM]` to indicate the direction for each dai. But `direction` is also used by SSP DAIs. See
[playback = info->direction[SNDRV_PCM_STREAM_PLAYBACK];
capture = info->direction[SNDRV_PCM_STREAM_CAPTURE];](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/intel/boards/sof_sdw.c#L1379)

File: sound/soc/intel/boards/sof_sdw.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > do we really want to do this? This would require us to have codec-specific stuff in the topology, no?

Maybe we can rename `codec_type` to `dai_type` in `sof_sdw_codec_info{}` and construct dai name with dai_type?
For example, `SDW0-Playback-amp`, ` SDW0-Playback-jack`

**Comment by bardliao** [comment]
> Not getting the introduction of the codec_name in the dai link name @bardliao. I am not sure why we would do this because there is currently no reference at all to the codecs in the topology

@plbossart This PR is a preparation of supporting rt712. rt712 supports 2 DAIs one for headset and one for amp. So that we need to create two dai links for them. The dai links could be:
```
sof_sdw sof_sdw: create dai link SDW0-Playback-jack, id 0     
sof_sdw sof_sdw: create dai link SDW0-Capture, id 1
sof_sdw sof_sdw: create dai link SDW0-Playback-amp, id 2
sof_sdw sof_sdw: create dai link SDW3-Capture, id 4
```

**Comment by plbossart** [comment]
@bardliao the topology and machine driver are related by ONE string: the dailink name. We cannot modify the existing dailink names otherwise we will break existing setups. we can add new dailinks for sure but the existing ones have to be left alone.

**Comment by bardliao** [comment]
> @bardliao the topology and machine driver are related by ONE string: the dailink name. We cannot modify the existing dailink names otherwise we will break existing setups. we can add new dailinks for sure but the existing ones have to be left alone.

Yes, that's why I check if the dailink name is duplicated first. There is no existing machine driver/topology supports more than one dailinks on the same sdw link with the same direction. In other words, if we find a duplicated dailink name, for example, a second `SDW0-Playback`, that must be a new configuration and we have to create a new topology for it. And I think we can edit the first `SDW0-Playback`'s name as well in this case since the topology is a new one.

File: sound/soc/intel/boards/sof_sdw_common.h
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_common.h]: > where does this come from?

Sorry, I forgot to answer this commit. The `3` is because I assume that we have at most 3 different functions, like jack, amp, and mic dais in a codec. Ideally, that should be allocated dynamically according to `dai_num`, but I didn't find a way to do so. Thus, I define a reasonable maxim dai number instead.

**Comment by bardliao** [comment]
> > @bardliao the topology and machine driver are related by ONE string: the dailink name. We cannot modify the existing dailink names otherwise we will break existing setups. we can add new dailinks for sure but the existing ones have to be left alone.
> 
> Yes, that's why I check if the dailink name is duplicated first. There is no existing machine driver/topology supports more than one dailinks on the same sdw link with the same direction. In other words, if we find a duplicated dailink name, for example, a second `SDW0-Playback`, that must be a new configuration and we have to create a new topology for it. And I think we can edit the first `SDW0-Playback`'s name as well in this case since the topology is a new one.

We need to find the codec info of the first duplicated dai link from codec_info_list. To keep it simpler, I will only append the dai type to the dai link name on the second duplicated dai link for now.

**Comment by plbossart** [comment]
SOFCI TEST

File: sound/soc/intel/boards/sof_sdw.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c:576]: Commit message: 

I don't think the dailink number if correct. We can have 3 types of dailinks, but we can have 5 dailinks already (jack in, jack out, amp out, amp feedback, smart mic).

I would also reword the last sentence as:

"The initial assumption if that we will create at most 3 dailink types for a codec, since this is the max known with upcoming SDCA devices. We may need to increase this number as new SDCA 'functions' become available.

One strong assumption is that all dailinks exposed are independent, as per SDCA directions.
"

File: sound/soc/intel/boards/sof_sdw.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: I don't get what the dais[0] is, and when/where we would consider the other dai descriptors?

**Comment by bardliao** [comment]
SOFCI TEST

**Comment by bardliao** [comment]
@plbossart Could you take a look at this PR?

File: sound/soc/intel/boards/sof_sdw.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: can those 0,1,4, etc be replaced by macros?

Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: commit message: you should clarify if there's any change in behavior or just the same results implemented differently.

File: sound/soc/intel/boards/sof_sdw.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: not following the [0] here.

**Review by plbossart — CHANGES_REQUESTED** [review]
not fully following this PR @bardliao

File: sound/soc/intel/boards/sof_sdw.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > I don't get what the dais[0] is, and when/where we would consider the other dai descriptors?

There is no function changed in the first commit. The other dai descriptors will be handled in the last commit.

File: sound/soc/intel/boards/sof_sdw.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > can those 0,1,4, etc be replaced by macros?

Yes, will do.

File: sound/soc/intel/boards/sof_sdw.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:576]: @plbossart The dailink number is tricky. I was trying to use flexible array, but I can't find a way to use flexible array here.
We do have 5 dailink types already, but we have directions in the sof_sdw_dai_info struct and create separated dailinks for playback and capture. So, we don't need to define different dai info arrays for different directions as long as the codec driver uses the same dai for playback and capture. As far as I know, most soundwire codec drivers use the same dai for playback and capture except the new Cirrus Logic codec driver.

File: sound/soc/intel/boards/sof_sdw.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > not following the [0] here.

We test [0] only is because all codec's dai_name in a dailink will be all the same.

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c]: that's not necessarily true. There could be a scenario where we have different amplifiers connected on the same dai, and one of the amplifiers deals with tweeter and the other one with woofer.

That's a big assumption here!

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > that's not necessarily true. There could be a scenario where we have different amplifiers connected on the same dai, and one of the amplifiers deals with tweeter and the other one with woofer.
> 
> That's a big assumption here!

Yes, but that's what we have today.

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw.c]: we need to plan for this to be changed. I am 100% sure someone will come up with designs with two different amps just to shave a bit of cost.

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > we need to plan for this to be changed. I am 100% sure someone will come up with designs with two different amps just to shave a bit of cost.

Agree, I modified the mc_dailink_exit_loop() function to check dais one by one. Verified with rt1316 + rt1318.

File: sound/soc/intel/boards/sof_sdw_common.h
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw_common.h]: Sorry @bardliao, is it intentional that 'unused' and 'jack_in' have the same ID? That seems wrong.

**Review by plbossart — COMMENTED** [review]
one open below @bardliao, looks good otherwise - to the extent that the machine driver code can be reviewed with a migraine, it's likely time to revisit all this and simplify or redo...

File: sound/soc/intel/boards/sof_sdw_common.h
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_common.h]: > Sorry @bardliao, is it intentional that 'unused' and 'jack_in' have the same ID? That seems wrong.

Yes, it was intentional . I use both 0 before using a macro. But, yes, it is better to use different number. Changed SDW_UNUSED_DAI_ID to -1 and added a check after getting dailink id now.

File: sound/soc/intel/boards/sof_sdw.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c]: is this correct?

the array is listed by direction, so the playback is first, then capture, but this shows the opposite order?

Not sure what I am missing.

File: sound/soc/intel/boards/sof_sdw.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw.c:631]: here the capture is listed second, so the jack above does seem wrong.

**Review by plbossart — CHANGES_REQUESTED** [review]
one big doubt on the JACK_IN/OUT order @bardliao

File: sound/soc/intel/boards/sof_sdw.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > is this correct?
> 
> the array is listed by direction, so the playback is first, then capture, but this shows the opposite order?
> 
> Not sure what I am missing.

Thank you @plbossart I fixed it now. It worked when I tested it because the defines were
#define SDW_JACK_IN_DAI_ID 0
#define SDW_JACK_OUT_DAI_ID 1

**Review by plbossart — DISMISSED** [review]
thanks @bardliao

File: sound/soc/intel/boards/sof_sdw.c
Comment (ranj063) [inline sound/soc/intel/boards/sof_sdw.c]: typo dsilink

File: sound/soc/intel/boards/sof_sdw.c
Comment (ranj063) [inline sound/soc/intel/boards/sof_sdw.c:570]: @bardliao can you please give an example of what the dai_link names become before and after this change?

File: sound/soc/intel/boards/sof_sdw.c
Comment (ranj063) [inline sound/soc/intel/boards/sof_sdw.c:1233]: i'd also split this par about adding codec_info into a separate patch first so the gist of this patch is easy to see

File: sound/soc/intel/boards/sof_sdw.c
Comment (ranj063) [inline sound/soc/intel/boards/sof_sdw.c]: I cant say that I understood exactly what's going on with the 4-level nested loop but it doesnt look pretty. I dont have an idea of how to make it better though

File: sound/soc/intel/boards/sof_sdw.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > typo dsilink

fixed

File: sound/soc/intel/boards/sof_sdw.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:1233]: > i'd also split this par about adding codec_info into a separate patch first so the gist of this patch is easy to see

ack

File: sound/soc/intel/boards/sof_sdw.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c:570]: > @bardliao can you please give an example of what the dai_link names become before and after this change?

For example, if a codec supports JACK on dai 0 and AMP on dai 1, the existing code will create dailinks
SDW0-Playback-SimpleJack or SDW0-Playback-SmartAmp for both dailinks,
and it will be SDW0-Playback-SimpleJack for dailink 0 and SDW0-Playback-SmartAmp for dailink 1 after this change.
I will add it to the commit message.

File: sound/soc/intel/boards/sof_sdw.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > I cant say that I understood exactly what's going on with the 4-level nested loop but it doesnt look pretty. I dont have an idea of how to make it better though

I added some comments in the code. Hope it helps.

Comment (ranj063) [reply sound/soc/intel/boards/sof_sdw.c]: @bardliao thanks then I have this suggestion and I think it makes it a lot easier to understand whats going on. Warning: Not even compile tested
```
typedef int (*codec_dai_exit_fn)(struct snd_soc_card *card, struct snd_soc_dai_link *dai_link);

/* helper to get the exit function callback for a codec DAI */
static codec_dai_exit_fn mc_find_codec_dai_exit_fn(struct snd_soc_card *card, struct snd_soc_dai_link_component *codec)
{
	int i, j;

	for (i = 0; i < ARRAY_SIZE(codec_info_list); i++) {
		for (j = 0; j < codec_info_list[i].dai_num; j++) {
			if (!strcmp(codec->dai_name, codec_info_list[i].dais[j].dai_name))
				return codec_info_list[i].dais[j].exit;
		}
	}
	
	return NULL;
}
	
static void mc_dailink_exit_loop(struct snd_soc_card *card)
{
	struct snd_soc_dai_link *link;
	int i, j;

	/* invoke exit callback for each codec in all links in the card */
	for_each_card_prelinks(card, i, link) {
		for (j = 0; j < link->num_codecs; j++) {
			/* 
			codec_dai_exit_fn exit = mc_find_codec_dai_exit_fn(card, &link->codecs[j]);
			if (exit) {
				exit(card, link);
                                break;
                        }
                }
	}
}
```

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: @ranj063 Thanks for the suggestion. I did slight change because we need to check codec_info_list one by one instead of checking each link. Because a codec dai could be use twice in the link, and we will call .exit function twice if we call .exit for each link.

**Review by ranj063 — APPROVED** [review]

File: sound/soc/intel/boards/sof_sdw.c
Comment (ranj063) [reply sound/soc/intel/boards/sof_sdw.c]: but thats why I had the break inside the if (exit). I think logically it makes to have the for_each_card_prelinks(card, i, link) as the outermost loop but I am not going to block the PR because of that

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw.c]: > but thats why I had the break inside the if (exit). I think logically it makes to have the for_each_card_prelinks(card, i, link) as the outermost loop but I am not going to block the PR because of that

The codec could be in different link, and we can have different codecs in the same link.
For example, 
SDW0-Plyaback uses rt711 for headset playback
SDW0-Captue uses rt711 for headset capture
SDW2-Playback uses rt1316 and rt1308 for speaker playback.
With the break inside the if(exit), rt711 .exit will be called twice, and rt1308 .exit will not be called.

**Review by plbossart — APPROVED** [review]

**Comment by plbossart** [comment]
SOFCI TEST

**Comment by plbossart** [comment]
no issues seen outside of usual suspend-resume TIMEOUTs, merging

**Comment by bardliao** [comment]
> no issues seen outside of usual suspend-resume TIMEOUTs, merging

@plbossart @ujfalusi Please don't submit this upstream for now. I need some fixup for SDW Pin index. Currently, the sdw cpu dai name will always be SDW Pin2 for playback and SDW Pin3 for capture. This doesn't work if we have 2 or more playback links or capture links. Sorry for not seeing this before merging.

**Comment by bardliao** [comment]
> > no issues seen outside of usual suspend-resume TIMEOUTs, merging
> 
> @plbossart @ujfalusi Please don't submit this upstream for now. I need some fixup for SDW Pin index. Currently, the sdw cpu dai name will always be SDW Pin2 for playback and SDW Pin3 for capture. This doesn't work if we have 2 or more playback links or capture links. Sorry for not seeing this before merging.

The fixup is https://github.com/thesofproject/linux/pull/4358. But I am not sure if this is a fixup or should land before `ASoC: Intel: sof_sdw: add multi dailink support for a codec`

**Pull request merged**
